File: go/problem-5/problem-5.go
package main

import (
    "fmt"
    "math"
    "../packages/utilities/mathutils"
)

func main() {
    firstTwenty := mathutils.NumbersBetween(1, 20)
    answer := lcmMultiple(firstTwenty)

    fmt.Println(answer)
}

func lcmMultiple(nums []int) int {
    var lcmCurrent int

    for i,num := range nums {
        if i == 0 {
            lcmCurrent = num
        } else {
            lcmCurrent = lcm(lcmCurrent, num)
        }
    }

    return lcmCurrent
}

func lcm(a int, b int) int {
    gcd := gcd(a, b)
    lcm := int(math.Abs(float64(a * b))) / gcd
    return lcm
}

func gcd(a int, b int) int {
    // Unlike in Java, we can't call this variable gcd. The compiler would think
    // we were referring to this variable when we call gcd() later.
    gcdValue := 1

    if (b == 0) {
        gcdValue = a
    } else if (a >= b && b > 0) {
        gcdValue = gcd(b, a % b)
    } else if (b >=a && a > 0) {
        gcdValue = gcd(b, a % b)
    }

    return gcdValue
}

File: java/problem-5/Problem5.java
import Utilities.math.MathUtilities;

public class Problem5 {
    public static void main(String[] args) {
        int[] numbersBelowTwenty = MathUtilities.numbersBetween(1,20);
        long answer = lcmMultiple(numbersBelowTwenty);
        System.out.println(answer);
    }

    public static long lcmMultiple(int[] nums) {
        long lcmCurrent = nums[0];

        for (int i = 1; i < nums.length; i++) {
            lcmCurrent = lcm(lcmCurrent, nums[i]);
        }

        return lcmCurrent;
    }

    public static long lcm(long a, long b) {
        long gcd = gcd(a, b);
        long lcm = Math.abs(a * b) / gcd;
        return lcm;
    }

    private static long gcd(long a, long b) {
        long gcd = 1;

        if (b == 0) {
            gcd = a;
        }
        else if (a >= b && b > 0) {
            gcd = gcd(b, a % b);
        }
        else if (b >= a && a > 0) {
            gcd = gcd(b, a % b);
        }

        return gcd;
    }
}
